### PR TITLE
fix: properly decode Path with whitespaces in ApplicationArchiveBuildStep.

### LINKED ISSUE #3034: Quarkus projects do not compile when project path contains whitespaces
**Describe the bug**
I am trying to build a quarkus project on our company jenkins. The path to the project in jenkins contains whitespaces.
"/var/lib/jenkins/jobs/wolf - project/workspace/project-web/target/classes"
The build fails.
If I remove these whitespaces, the build is successfull.

**Expected behavior**
The project builds without errors caused by internal quarkus maven plugin problems.

**Actual behavior**
Build fails with the following error:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:0.18.0:build (default) on project documentrepository-web: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR] [error]: Build step io.quarkus.deployment.index.ApplicationArchiveBuildStep#build threw an exception: java.lang.RuntimeException: java.io.FileNotFoundException: /var/lib/jenkins/jobs/wolf%20-%20project/workspace/project-web/target/classes (Datei oder Verzeichnis nicht gefunden)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :project-web
```

**To Reproduce**
1. Move an existing, successfully compiling quarkus project into a path with whitespaces
2. Compile
3. Does not compile

**Additional context**
quarkus 018.0
java 8
Maven 3.6.1

### PR DESCRIPTION
**Pull request by machi1990**
@Postremus can you confirm if this PR fixes #3034? 

Thanks.

/cc @gsmet @Postremus

### CODE REVIEW

**Comment by Postremus** [comment]
@machi1990 I will have time to check it this evening (CET)

**Comment by machi1990** [comment]
Thanks @Postremus. You can clone my branch and build the project using ‘mvn clean install -DskipTests -DskipITs’. You’ll have a 999-SNAPSHOT Quarkus version which you can use.

File: core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java [hunk 1]
```diff
@@ -173,7 +173,7 @@ static Path urlToPath(URL url) {
             } else if (url.getProtocol().equals("file")) {
                 int index = url.getPath().lastIndexOf("/META-INF");
```
Comment (Postremus) [inline core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: url.toURI()

File: core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java [hunk 2]
```diff
@@ -173,7 +173,7 @@ static Path urlToPath(URL url) {
...
                 if (index == -1) {
-                    return Paths.get(url.getPath());
+                    return Paths.get(url.toURI().getPath());
                 }
                 String pathString = url.getPath().substring(0, index);
```
Comment (Postremus) [inline core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: url.toURI()

**Review by Postremus — COMMENTED** [review]
see comments for additional required changes

**Comment by Postremus** [comment]
@machi1990 
It is working now, but the ApplicationArchiveBuildStep needed some small additional modifications. I provided them via a review.

File: core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java (hunk 2 above)
Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: @machi1990 maybe let's create a variable for the `Path` once and for all? :)

Comment (machi1990) [reply core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: Thanks @Postremus for the review and confirmation.  I’ll revisit this and add more tests. 

@gsmet This is what I thought when looking at the codebase. I wanted to be sure if it fixes the issue above before. I’ll push my final changes later on.

Comment (machi1990) [reply core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: Done

**Comment by machi1990** [comment]
> @machi1990
> It is working now, but the ApplicationArchiveBuildStep needed some small additional modifications. I provided them via a review.

Thanks for the review @Postremus and for checking out. It was a bit tricky to reproduce locally. But finally the fix is here much thanks to your help.

**Comment by machi1990** [comment]
@gsmet ping.

File: core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java [hunk 3]
```diff
@@ -171,13 +172,12 @@ static Path urlToPath(URL url) {
                 String jarPath = url.getPath().substring(0, url.getPath().lastIndexOf('!'));
                 return Paths.get(new URI(jarPath));
             } else if (url.getProtocol().equals("file")) {
-                int index = url.getPath().lastIndexOf("/META-INF");
+                final String path = Paths.get(url.toURI()).toString();
+                final int index = path.lastIndexOf(File.separator + "META-INF");
                 if (index == -1) {
-                    return Paths.get(url.getPath());
+                    return Paths.get(path);
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: We could avoid recreating the path here, maybe?

File: core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java [hunk 4]
```diff
@@ -171,13 +172,12 @@ static Path urlToPath(URL url) {
...
                 }
-                String pathString = url.getPath().substring(0, index);
-                Path path = Paths.get(new URI(url.getProtocol(), url.getHost(), pathString, null));
-                return path;
+                return Paths.get(path.substring(0, index));
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: Not sure I understand why the code was initially like that.

File: core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java (hunk 3 above)
Comment (machi1990) [reply core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java]: Done

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Merged, thanks!

**Pull request merged**